---
title: "BCG case"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
#library(MPOInt)
library(dplyr)
library(clusterProfiler)
library(enrichplot)
library(stringr)
library(tibble)
library(disgenet2r)
library(ggvenn)
library(reticulate)
library(igraph)
```

```{r}
total_p_threshold <- 0.05
total_FC_threshold <- 0.5
phospho_p_threshold <- 0.05
phospho_FC_threshold <- 0.5
histone_p_threshold <- 0.05
histone_FC_threshold <- 0.5
```

```{r d2r-setup, eval = FALSE}
api_key <- "#####"
Sys.setenv(DISGENET_API_KEY= api_key)
```

```{r python-1st-setup, eval = FALSE}
virtualenv_create("r-reticulate")
use_virtualenv("r-reticulate")
virtualenv_install("r-reticulate", c("pandas", "networkx", "pyvis"), action = "add")
```
# 1. Proteome data processing

```{r Total processing}
data(THP1_total)
# Renaming to reader-friendly names
THP1_total <- dplyr::rename(THP1_total, all_of(c(
  "log2(FC)" = "Abundance Ratio (log2): (JMI) / (Control)",
  "P value" = "Abundance Ratio Adj. P-Value: (JMI) / (Control)"
)))
# Mapping ENTREZ IDs and a gene symbol column to UniProt IDs for later analysis
THP1_total <- dplyr::left_join(THP1_total, OrganismDbi::select(
  Homo.sapiens::Homo.sapiens,
  keys = OrganismDbi::keys(Homo.sapiens::Homo.sapiens, keytype = "UNIPROT"),
  columns = c("ENTREZID", "SYMBOL"), keytype = "UNIPROT"
), by = dplyr::join_by(Accession == UNIPROT), keep = FALSE)
# Dropping anything with an undefined or duplicate symbol
THP1_total <- THP1_total %>%
  tidyr::drop_na(SYMBOL) %>%
  dplyr::distinct(SYMBOL, .keep_all = TRUE)
# Assigning row names
THP1_total <- THP1_total %>%
  dplyr::mutate("SYMBOL2" = `SYMBOL`) %>%
  tibble::column_to_rownames(var = "SYMBOL")

# Creating a separate set of significant points for later analysis
THP1_total_sig <- dplyr::filter(THP1_total, `P value` < total_p_threshold)
```

## 1.1. Volcano plots
```{r}
# Setting this value - geom_text_repel will only show so many labels at once
# for legibility reasons. Adjust this threshold to show more/fewer points.
options("ggrepel.max.overlaps" = 20)

volcanoplot(df = THP1_total, fc_col = "log2(FC)",
            p_col = "P value", label_col = "SYMBOL2") +
  ggtitle("Volcano plot") + xlab("log2(FC)") + ylab("-log10(P-value)")
```

```{r Groups}
# For known groups of interest
volcano_groups(df = THP1_total, fc_col = "log2(FC)", p_col = "P value", label_col = "SYMBOL2",
  group_list = list("Group 1" = c("MORF4L1", "SAMSN1", "DTX3L", "MBD3",  "HSF1",
  "EYA3", "NCAPD2", "TAF9", "PPP4C", "ZNF638"), "Group 2" = c("CD9", "CD82", "NFKB1", "BST2", "CTSH"))) +
  ggtitle("Grouped volcano plot") + xlab("log2FC") + ylab("-log10P")
# TODO fix order of points/colors
```

## 1.2. GSEA
```{r}
total_GSEA_list <- THP1_total$`log2(FC)`
names(total_GSEA_list) <- THP1_total$ENTREZID
total_GSEA_list <- na.omit(total_GSEA_list)
total_GSEA_list <- sort(total_GSEA_list, decreasing = TRUE)
```

```{r}
total_GSEA_GO <- gseGO(geneList = total_GSEA_list, OrgDb = "org.Hs.eg.db",
                    verbose = FALSE, ont = "ALL")
total_GSEA_KEGG <- gseKEGG(geneList = total_GSEA_list, verbose = FALSE)
```

```{r, fig.height = 6, fig.width = 6, fig.dpi = 300}
if (nrow(total_GSEA_GO@result) > 0) {
  dotplot(total_GSEA_GO, showCategory = 20, font.size = rel(1), label_format = 40, title = "GO enrichment",
          split = "ONTOLOGY") + facet_wrap(ONTOLOGY~.sign)
} else {
  print("Error: no terms enriched under P-value cutoff (total_GSEA_GO)")
}

if (nrow(total_GSEA_KEGG@result) > 0) {
  dotplot(total_GSEA_KEGG, showCategory = 20, font.size = rel(1), label_format = 40, title = "KEGG enrichment",
          split = ".sign") + facet_wrap(.~.sign)
} else {
  print("Error: no terms enriched under P-value cutoff (total_GSEA_KEGG)")
}
```
```{r clear-1}
rm(THP1_total, total_GSEA_GO, total_GSEA_KEGG, total_GSEA_list)
```



# 2. Phosphoproteome data processing

```{r Phosphoproteome processing}
# Renaming to reader-friendly names not necessary here
data(THP1_phospho)
# Mapping ENTREZ IDs and a gene symbol column to UniProt IDs for later analysis
THP1_phospho <- dplyr::left_join(THP1_phospho, OrganismDbi::select(
  Homo.sapiens::Homo.sapiens,
  keys = OrganismDbi::keys(Homo.sapiens::Homo.sapiens, keytype = "ENTREZID"),
  columns = c("UNIPROT", "SYMBOL"), keytype = "ENTREZID"
), by = dplyr::join_by(SYMBOL2 == SYMBOL, ENTREZID), keep = FALSE, relationship = "many-to-many")
# Dropping anything with an undefined or duplicate symbol
THP1_phospho <- THP1_phospho %>%
  tidyr::drop_na(SYMBOL2) %>%
  dplyr::distinct(SYMBOL2, .keep_all = TRUE)
# Assigning row names
THP1_phospho <- THP1_phospho %>%
  dplyr::mutate("SYMBOL" = `SYMBOL2`) %>%
  tibble::column_to_rownames(var = "SYMBOL")

# Creating a separate set of significant points for later analysis
THP1_phospho_sig <- dplyr::filter(THP1_phospho, `P value` < phospho_p_threshold)
```

## 2.1. Volcano plots
```{r}
# Setting this value - geom_text_repel will only show so many labels at once
# for legibility reasons. Adjust this threshold to show more/fewer points.
options("ggrepel.max.overlaps" = 20)

volcanoplot(df = THP1_phospho, fc_col = "log2(FC)",
            p_col = "P value", label_col = "SYMBOL2") +
  ggtitle("Volcano plot") + xlab("log2(FC)") + ylab("-log10(P-value)")
```

```{r Groups}
# For known groups of interest
volcano_groups(df = THP1_phospho, fc_col = "log2(FC)", p_col = "P value", label_col = "SYMBOL2",
  group_list = list("Group 1" = c("ASH2L", "ASXL1", "ASXL2", "BAP1", "BMI1", "CHD4", "EZH2", "DMAP1",
  "EP400", "GATAD2B", "KANSL1", "KANSL3", "KMT2C", "KMT2D", "MBD3",
  "MEN1", "MGA", "MRGBP", "MTA1", "NCOA6", "PHC2", "PHF20L1", "REST",
  "RING1", "RUVBL1", "SAP130", "SCML2", "SIRT3", "SIRT6", "EP300"), "Group 2" = c("FANCI", "CDK14", "PRDM1", "ICAM1", "CDK1", "RUNX3", "NOTCH3", "CASP8",
  "BST2", "CTSL", "MX2", "CASP1", "STAT2", "MX1"))) +
  ggtitle("Grouped volcano plot") + xlab("log2FC") + ylab("-log10P")
# TODO fix order of points/colors
```

## 2.2. GSEA
```{r}
phospho_GSEA_list <- THP1_phospho$`log2(FC)`
names(phospho_GSEA_list) <- THP1_phospho$ENTREZID
phospho_GSEA_list <- na.omit(phospho_GSEA_list)
phospho_GSEA_list <- sort(phospho_GSEA_list, decreasing = TRUE)
```

```{r}
phospho_GSEA_GO <- gseGO(geneList = phospho_GSEA_list, OrgDb = "org.Hs.eg.db",
                    verbose = FALSE, ont = "ALL")
# By visual inspection of graphs, this contained redundant terms
simple_phospho_GSEA_GO <- clusterProfiler::simplify(phospho_GSEA_GO)
phospho_GSEA_KEGG <- gseKEGG(geneList = phospho_GSEA_list, verbose = FALSE)
```

```{r}
if (nrow(simple_phospho_GSEA_GO@result) > 0) {
  dotplot(simple_phospho_GSEA_GO, showCategory = 20, font.size = rel(1), label_format = 50, title = "GO enrichment",
          split = "ONTOLOGY") + facet_wrap(ONTOLOGY~.sign, scales = "free_y", ncol = 2)
} else {
  print("Error: no terms enriched under P-value cutoff (simple_phospho_GSEA_GO)")
}

if (nrow(phospho_GSEA_KEGG@result) > 0) {
  dotplot(phospho_GSEA_KEGG, showCategory = 20, font.size = rel(1), label_format = 40, title = "KEGG enrichment",
          split = ".sign") + facet_wrap(.~.sign, scales = "free", ncol = 2)
} else {
  print("Error: no terms enriched under P-value cutoff (phospho_GSEA_KEGG)")
}
```

```{r clear-2}
rm(phospho_GSEA_GO, phospho_GSEA_KEGG, simple_phospho_GSEA_GO, THP1_phospho,
   phospho_GSEA_list)
```

# 3. Histone processing
```{r fragment-histone-processing, eval = TRUE}
data(THP1_histones)
histone_frag_raw <- tidyr::drop_na(THP1_histones)
histone_frag_raw$PTM <- histone_frag_raw$PTM %>% 
  str_replace_all("H3_3_8", "H3 3-8") %>% 
  str_replace_all("H3_9_17", "H3 9-17") %>% 
  str_replace_all("H3_18_26", "H3 18-26") %>% 
  str_replace_all("H3_27_40", "H3 27-40") %>% 
  str_replace_all("H33_27_40", "H3.3 27-40") %>% 
  str_replace_all("H3_41_49", "H3 41-49") %>% 
  str_replace_all("H3_54_63", "H3 54-63") %>% 
  str_replace_all("H3_73_83", "H3 73-83") %>% 
  str_replace_all("H3_117_128", "H3 117-128") %>% 
  str_replace_all("H4_4_17", "H4 4-17") %>% 
  str_replace_all("H4_20_23", "H4 20-23") %>% 
  str_replace_all("H4_24_35", "H4 24-35") %>% 
  str_replace_all("H4_40_45", "H4 40-45") %>% 
  str_replace_all("H4_79_92", "H4 79-92") %>% 
  str_replace_all("H1_1_35", "H1 1-35") %>% 
  str_replace_all("H1_54_81", "H1 54-71") %>% 
  str_replace_all("H2A1_36_42", "H2A1 36-42") %>% 
  str_replace_all("H2A3_36_42", "H2A3 36-42") %>% 
  str_replace_all("H2AX_36_42", "H2AX 36-42") %>% 
  str_replace_all("H2A1_4_11", "H2A1 4-11") %>% 
  str_replace_all("H2AJ_4_11", "H2AJ 4-11") %>% 
  str_replace_all("H2AX_4_11", "H2AX 4-11") %>% 
  str_replace_all("H2A1_1_11", "H2A1 1-11") %>% 
  str_replace_all("H2AV_1_19", "H2AV 1-19") %>% 
  str_replace_all("H2AZ_1_19", "H2AZ 1-19") %>% 
  str_replace_all("H2A1_12_17", "H2A1 12-17") %>% 
  str_replace_all("H2A3_12_17", "H2A3 12-17") %>% 
  str_replace_all("H2A1_72_77", "H2A1 72-77") %>% 
  str_replace_all("H2A_82_88", "H2A 82-88") %>% 
  str_replace_all("H2A_1_88", "H2A 1-88") %>% 
  str_replace_all("H2B_1_29", "H2B 1-29")

```

**For summary statistics data**: switch all chunks in section 3 with `eval = FALSE` in the chunk header to `eval = TRUE` and vice versa. Do not alter chunk headers without this item.
```{r histone-import-mean, eval = FALSE}
histone_frag_raw <- data(MCF7_histones)
histone_frag_raw <- histone_frag_raw %>% 
  dplyr::select(c("PTM", "MCF7 ave (5 reps)", "MCF7 std", "total ave", "avg std")) %>% 
  mutate("log2(FC)" = log2(`MCF7 ave (5 reps)`/`total ave`)) %>% 
  mutate("T score" = NA) %>% 
  mutate("P value" = NA)
histone_frag <- histone_frag_raw
histone_frag <- histone_frag %>%
  mutate("T score" = (`MCF7 ave (5 reps)` - `total ave`)/(`MCF7 std`/sqrt(5))) %>% 
  mutate("P value" = pt(q = `T score`, df = 4))
```

## 3.1. Individual values
### 3.1.1. Fragments
```{r fragment-histone-processing2, eval = TRUE}
# Group two conditions together
histone_frag <- histone_frag_raw
histone_frag <- histone_frag %>%
  relocate(contains("I"), .after = last_col()) %>% 
  relocate("PTM")

# Calculate infected/control fold change and P value
histone_frag <- histone_frag %>% 
  rowwise() %>% 
  mutate(c_mean = mean(c_across("C1":"C5")),
         i_mean = mean(c_across("I1":"I5")),
         FC = c_across(i_mean)/c_across(c_mean),
         "log2(FC)" = log2(FC),
         is_constant = all(c_across("C1":"C5") == c_across("I1":"I5")))

histone_frag <- column_to_rownames(histone_frag, var = "PTM")
histone_frag$PTM <- rownames(histone_frag)

histone_frag$"P value" <- NA
histone_p <- histone_frag %>% 
  rowwise() %>% 
  filter(is_constant == FALSE) %>% 
  # Removing any proteins that have all equal values, which gives an 
  # error when running t.test()
  mutate("P value" = t.test(c_across("C1":"C5"), c_across("I1":"I5"),
         alternative = "two.sided", var.equal = TRUE)$p.value)

histone_frag <- full_join(histone_frag, histone_p)

histone_frag$`log2(FC)`[histone_frag$`log2(FC)` == "NaN"] <- NA
histone_frag$FC[histone_frag$FC == "NaN"] <- NA
```

```{r}
# Setting this value - geom_text_repel will only show so many labels at once
# for legibility reasons. Adjust this threshold to show more/fewer points.
options("ggrepel.max.overlaps" = 20)

volcano_epi(df = histone_frag, fc_col = "log2(FC)",
            p_col = "P value", label_col = "PTM") +
  labs(title = "Volcano plot", x = "log2(FC)", y = "-log10(P-value)")
```


### 3.1.2. Single
```{r warning = FALSE}
histone_single_raw <- histone_frag_raw
histone_single_raw <- histone_single_raw[!grepl("\\.", histone_single_raw$PTM),]
histone_single_unmod <- histone_single_raw[grepl("unmod", histone_single_raw$PTM),]
histone_single_raw <- histone_single_raw[!grepl("unmod", histone_single_raw$PTM),]

#TODO silence warnings
histone_single_raw <- histone_single_raw %>% 
  mutate("subset" = str_extract(`PTM`, "([^\ ]+$)")) %>% 
  mutate("PTM4" = str_split_i(`subset`, "(?=K)", i = -1)) %>% 
  mutate("PTM3" = str_split_i(`subset`, "(?=K)", i = -2)) %>% 
  mutate("PTM2" = str_split_i(`subset`, "(?=K)", i = -3)) %>% 
  # add line above for max co-occurring PTMs, then remove
  mutate("PTM1" = str_remove(`subset`, `PTM2`)) %>% 
  mutate("PTM1" = str_remove(`PTM1`, `PTM3`)) %>% 
  mutate("PTM1" = str_remove(`PTM1`, `PTM4`))

histone_single_raw <- replace(histone_single_raw, histone_single_raw == "", NA)

histone_single_long <- histone_single_raw %>% 
  pivot_longer(cols = c("PTM4", "PTM3", "PTM2", "PTM1"), names_to = NULL,
               values_to = "single_PTM", values_drop_na = TRUE) %>% 
  mutate("PTM" = str_remove(`PTM`, `subset`)) %>% 
  unite(col = "PTM", c(`PTM`, `single_PTM`), sep = " ", remove = TRUE) %>% 
  mutate("PTM" = str_squish(PTM))


histone_single_long$PTM <- str_remove(histone_single_long$PTM, " .+ ") %>% 
  str_replace("me1", "me")
histone_single_long <- bind_rows(histone_single_long, histone_single_unmod)
```
If the data being analyzed has standard deviation and mean listed but not individual values, calculate the T score and then the P value from that. To run this chunk, remove `eval = FALSE` in the chunk header and add it to the `single-stats-p` chunk.



```{r single-stats-t, eval = FALSE}
histone_single_raw <- histone_single_long %>% 
  group_by(`PTM`) %>% 
  summarize("sample ave" = mean(`MCF7 ave (5 reps)`),
            "sample std" = mean(`MCF7 std`),
            "total ave" = mean(`total ave`),
            "avg std" = mean(`avg std`),
            "T score" = "",
            "P value" = "") %>% 
  mutate("log2(FC)" = log2(`sample ave`/`total ave`))

histone_single_p <- histone_single_raw %>% 
  mutate("T score" = (`sample ave` - `total ave`)/(`sample std`/sqrt(5))) %>% 
  mutate("P value" = pt(q = `T score`, df = 4))

histone_single_p$PTM <- str_remove(histone_single_p$PTM, " .+ ") %>% 
  str_remove("un") %>% 
  str_replace("me1", "me")
```

```{r single-stats-p, eval = TRUE}

histone_single_long_p <- histone_single_long %>% 
  group_by(`PTM`) %>% 
  summarize("C1" = mean(C1), "C2" = mean(C2), "C3" = mean(C3),
            "C4" = mean(C4), "C5" = mean(C5), "I1" = mean(I1),
            "I2" = mean(I2), "I3" = mean(I3), "I4" = mean(I4),
            "I5" = mean(I5))

# Calculate infected/control fold change and P value
histone_single_p <- histone_single_long_p %>% 
  rowwise() %>% 
  mutate(c_mean = mean(c_across("C1":"C5")),
         i_mean = mean(c_across("I1":"I5")),
         FC = c_across(i_mean)/c_across(c_mean),
         "log2(FC)" = log2(FC),
         is_constant = all(c_across("C1":"C5") == c_across("I1":"I5")))

histone_single_p <- column_to_rownames(histone_single_p, var = "PTM")
histone_single_p$PTM <- rownames(histone_single_p)

histone_single_p$"P value" <- NA
single_p <- histone_single_p %>% 
  rowwise() %>% 
  filter(is_constant == FALSE) %>% 
  # Removing any proteins that have all equal values, which gives an 
  # error when running t.test()
  mutate("P value" = t.test(c_across("C1":"C5"), c_across("I1":"I5"),
         alternative = "two.sided", var.equal = TRUE)$p.value)

histone_single_p <- full_join(histone_single_p, single_p)

histone_single_p$`log2(FC)`[histone_single_p$`log2(FC)` == "NaN"] <- NA
histone_single_p$FC[histone_single_p$FC == "NaN"] <- NA
```

```{r}
# Setting this value - geom_text_repel will only show so many labels at once
# for legibility reasons. Adjust this threshold to show more/fewer points.
options("ggrepel.max.overlaps" = 20)

volcano_epi(df = histone_single_p, fc_col = "log2(FC)",
            p_col = "P value", label_col = "PTM") +
  labs(title = "Volcano plot", x = "log2(FC)", y = "-log10(P-value)")
```
```{r clear-3}
rm(THP1_histones, histone_frag, histone_frag_raw, histone_p,
   histone_single_long, histone_single_long_p, histone_single_raw,
   histone_single_unmod, single_p)
```

# 4. Disease enrichment

```{r GDA-CUI-fetch}
GDA_table_BlCa <- get_umls_from_vocabulary(disease = "bladder cancer",  vocabulary = "NAME" ,  limit = 10)@qresult
knitr::kable(GDA_table_BlCa, caption = "List of CUIs that map to bladder cancer", row.names = F) 
```

```{r GDA-setup}
GDA_BlCa <- disease2gene(disease = "UMLS_C0005684", database = "CURATED")@qresult
GDA_melanoma <- disease2gene(disease = "UMLS_C0025202", database = "CURATED")@qresult
GDA_RSV <- disease2gene(disease = "UMLS_C0035235", database = "CURATED")@qresult
```


```{r GDA, fig.asp = 1}
# Get lists of proteins that were significant AND above the FC threshold
total_FC <- filter(THP1_total_sig, abs(`log2(FC)`) > total_FC_threshold)
phospho_FC <- filter(THP1_phospho_sig, abs(`log2(FC)`) > phospho_FC_threshold)

# Generate plot
ggvenn(list(
  "NMIBC" = GDA_BlCa$gene_symbol,
  "Total proteome" = total_FC$SYMBOL2,
  "Phosphoproteome" = phospho_FC$SYMBOL2
)) + ggtitle("Figure 6a: Venn diagram") + theme(
  plot.title.position = "plot",
  plot.title = element_text(hjust = 0.5))

ggvenn(list(
  "Melanoma" = GDA_melanoma$gene_symbol,
  "Total proteome" = total_FC$SYMBOL2,
  "Phosphoproteome" = phospho_FC$SYMBOL2
)) + ggtitle("Figure 6a: Venn diagram") + theme(
  plot.title.position = "plot",
  plot.title = element_text(hjust = 0.5))

ggvenn(list(
  "RSV" = GDA_RSV$gene_symbol,
  "Total proteome" = total_FC$SYMBOL2,
  "Phosphoproteome" = phospho_FC$SYMBOL2
)) + ggtitle("Figure 6a: Venn diagram") + theme(
  plot.title.position = "plot",
  plot.title = element_text(hjust = 0.5))
```
```{r GDA-list, results = "asis"}
# Print the list of proteins in each intersection
cat(paste("**Disease/Phospho:** ", paste(intersect(GDA_BlCa$gene_symbol, phospho_FC$SYMBOL2), collapse = ", ")))
cat(paste("**Disease/Total:** ", paste(intersect(GDA_BlCa$gene_symbol, total_FC$SYMBOL2), collapse = ", ")))
cat(paste("**Total/Phospho:** ", paste(intersect(total_FC$SYMBOL2, phospho_FC$SYMBOL2), collapse = ", ")))
cat(paste("**Disease/Phospho/Total:** ", paste(intersect(intersect(GDA_BlCa$gene_symbol, phospho_FC$SYMBOL2), total_FC$SYMBOL2), collapse = ", ")))
```

```{r clear-4}
rm(GDA_BlCa, GDA_melanoma, GDA_RSV, GDA_table_BlCa)
```


# 5. Epigenetic network
This section develops a network visualization of all epigenetic interactions as defined by EpiFactors (https://epifactors.autosome.org/). Increased dataset size may make networks difficult to visualize, so an interactive network is drawn in Python and rendered as an embedded HTML file.

```{r network-setup}
data(genes)
total_join <- inner_join(genes, dplyr::select(THP1_total_sig, "log2(FC)", "P value", "SYMBOL2"), by = join_by("HGNC approved symbol" == "SYMBOL2"), multiple = "any") %>% 
  mutate("Source" = "Total")


phospho_join <- inner_join(genes, dplyr::select(THP1_phospho_sig, "log2(FC)", "P value", "SYMBOL2"), by = join_by("HGNC approved symbol" == "SYMBOL2"), multiple = "any") %>% 
  mutate("Source" = "Phospho")

joined_data <- full_join(total_join, phospho_join)
joined_data$Diff <- case_when(joined_data$`log2(FC)` > 0 ~ "Up",
                             joined_data$`log2(FC)` < 0 ~ "Down")
```

```{r network-proteins}
# Plotting all relevant proteins from the two combined datasets
ggplot(joined_data, aes(x = `log2(FC)`, y = -log10(`P value`))) +
  geom_point(aes(color = Source)) +
  geom_vline(xintercept = c(total_FC_threshold, -total_FC_threshold),
             alpha = 0.5, linetype = "dashed") +
  geom_hline(yintercept = -log10(total_p_threshold),
             alpha = 0.5, linetype = "dashed") +
  labs(x = "log2(Fold Change)", y = "-log10(P value)",
       title = "Figure 7a: Volcano plot") +
  geom_text_repel(aes(label = `HGNC approved symbol`), show.legend = FALSE,
                  size = rel(3), max.overlaps = 20) +
  theme(legend.position = "bottom")
```
```{r network-histone-import}
# Formatting histone targets
targets <- str_replace_all(str_flatten_comma(str_unique(c(joined_data$Product, joined_data$`Target entity`))), ", ", "|")

# filter by that list
total_histone_int <- filter(histone_single_p, grepl(targets, PTM))
total_histone_int$Diff <- case_when(total_histone_int$`log2(FC)` > 0 ~ "Up",
                             total_histone_int$`log2(FC)` < 0 ~ "Down",
                             .default = "0")
total_histone_int <- filter(total_histone_int, !grepl("unmod", PTM))

# Making sure that no histone data has gotten lost in processing - 
# this should match the figure from the earlier section
ggplot(total_histone_int, aes(x = `log2(FC)`, y = -log10(`P value`))) +
  geom_point() +
  geom_vline(xintercept = c(total_FC_threshold, -total_FC_threshold),
             alpha = 0.5, linetype = "dashed") +
  geom_hline(yintercept = -log10(total_p_threshold),
             alpha = 0.5, linetype = "dashed") +
  labs(x = "log2(Fold Change)", y = "-log10(P value)",
       title = "Figure 7b: Volcano plot") +
  geom_text_repel(aes(label = PTM), show.legend = FALSE, size = rel(3)) +
  theme(legend.position = "bottom")
```

```{r network-genes}
# Pivoting and formatting the gene list
genes_pivoted <- dplyr::select(genes, c("HGNC approved symbol", "Protein complex", "Target entity", "Product")) %>% dplyr::rename("Symbol" = "HGNC approved symbol", "Complex" = "Protein complex", "Target" = "Target entity", "Product" = "Product")

genes_pivoted <- genes_pivoted %>% 
  separate_wider_delim(cols = c("Complex", "Target", "Product"), delim = ", ",
                       names_sep = "_", too_few = "align_start")

genes_pivoted <- genes_pivoted %>% 
  pivot_longer(cols = dplyr::starts_with("Complex"), names_to = NULL,
               values_to = "Complex", values_drop_na = TRUE) %>% 
  pivot_longer(cols = dplyr::starts_with("Target"), names_to = NULL,
               values_to = "Target", values_drop_na = TRUE) %>% 
  pivot_longer(cols = dplyr::starts_with("Product"), names_to = NULL,
               values_to = "Product", values_drop_na = TRUE)


genes_data_joined <- inner_join(genes_pivoted,
                                dplyr::select(joined_data, c("HGNC approved symbol",
                                                      "log2(FC)", "P value", "Source")),
                                by = join_by("Symbol" == "HGNC approved symbol"),
                                relationship = "many-to-many")
```

```{r network-format}
# Getting the lists into the right format for the igraph visualization
genes_histone_pivoted <- pivot_longer(genes_data_joined, cols = c("Target", "Product"), names_to = NULL,
             values_to = "PTM") %>% distinct()
genes_histone_pivoted$Diff <- case_when(genes_histone_pivoted$`log2(FC)` > 0 ~ "Up",
                             genes_histone_pivoted$`log2(FC)` < 0 ~ "Down",
                             .default = "0")

```


```{r}
genes_histone_joined <- right_join(genes_histone_pivoted,
                                dplyr::select(total_histone_int, c("PTM",
                                                      "log2(FC)", "P value", "Diff")),
                                by = c("PTM" = "PTM"), suffix = c("_gene", "_hist"),
                                relationship = "many-to-many"
                                )
genes_histone_joined <- relocate(genes_histone_joined, Complex, .after = PTM)
genes_histone_joined[genes_histone_joined == "#"] <- "No ID"
genes_histone_joined$Complex <- genes_histone_joined$Complex %>%  replace_na("No ID")
```


```{r}
#new_vertices <- pivot_longer(genes_histone_joined, cols = c("Diff_gene", "Diff_hist"), names_to = NULL, values_to = "Diff")
```




```{r}
complexes <- unique(genes_histone_joined$Complex)
n_complexes <- seq(1, length(complexes))
color_opts <- c("#FFFFFF", qualpalr::qualpal(n = (length(complexes) - 1), colorspace = "rainbow")$hex)
complex_df <- data.frame(complexes, n_complexes, color = color_opts)
```


```{r}
histone_df <- dplyr::select(total_histone_int, c("PTM", "Diff"))
histone_df$Symbol <- histone_df$PTM
histone_df$Complex <- "No ID"
histone_df$Source <- "Histone"

new_df <- bind_rows(histone_df, dplyr::select(genes_histone_pivoted, !c("log2(FC)", "P value"))) %>% relocate("Symbol")
new_df <- filter(new_df, grepl("^H", PTM))
new_df$Complex <- case_when(new_df$Complex == "#" ~ "No ID",
                            .default = new_df$Complex)
```





```{r}
complex_vertices <- data.frame(vertices = c(unique(genes_histone_joined$Symbol), unique(genes_histone_joined$PTM)))

network_df <- full_join(genes_histone_joined, complex_df, by = join_by("Complex" == "complexes")) %>% relocate(c("Symbol", "PTM"))

network_df <- left_join(complex_vertices, network_df, by = join_by("vertices" == "Symbol")) %>% rename("Symbol" = "vertices")

network_df$Symbol <- case_when(is.na(network_df$Symbol) ~ network_df$PTM,
                               .default = network_df$Symbol)
network_df$PTM <- case_when(is.na(network_df$PTM) ~ network_df$Symbol,
                               .default = network_df$PTM)
network_df$Source <- case_when(is.na(network_df$Source) ~ "Histone",
                               .default = network_df$Source)
network_df$Complex <- case_when(is.na(network_df$Complex) ~ "No ID",
                               .default = network_df$Complex)
network_df$Diff <- case_when(network_df$Source == "Histone" ~ network_df$Diff_hist,
                             .default = network_df$Diff_gene)
network_df <- drop_na(network_df, Diff)

complex_vert_attr <- left_join(complex_vertices, network_df, by = join_by("vertices" == "Symbol"), multiple = "all") %>% dplyr::select(c("vertices", "Complex", "PTM", "Source", "Diff")) %>% distinct(vertices, .keep_all = TRUE) %>% drop_na(vertices)

complex_graph <- graph_from_data_frame(d = network_df, vertices = complex_vert_attr, directed = FALSE)
```

These alternative visualizations show the members of the identified complexes without connections.

```{r network-bars}
with_id <- filter(network_df, Complex != "No ID")


sig_network <- distinct(group_by(with_id, Complex), Symbol, .keep_all = TRUE)
sig_network["weight"] <- 1
sig_network$Complex <- factor(sig_network$Complex, levels = unique(as.character(sig_network$Complex)))
```

```{r network-bins}
ggplot(sig_network, aes(x = Complex, y = Symbol, fill = PTM)) +
  geom_bin_2d() +
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1),
        axis.title.x = element_blank(), axis.title.y = element_blank()) +
  paletteer::scale_fill_paletteer_d("ggsci::default_igv") +
  scale_y_discrete(limits = rev) +
  labs(title = "Figure 7d: Network bins")
```

## 5.1 Python visualization

```{r py-import}
new_df$Shapes <- case_when(new_df$Diff == "Up" ~ "triangle",
                             new_df$Diff == "Down" ~ "triangleDown",
                          new_df$Diff == "0" ~ "ellipse")
new_df$Colors <- case_when(new_df$Source == "Total" ~ "red",
                          new_df$Source == "Phospho" ~ "blue",
                          new_df$Source == "Histone" ~ "gray")

sig_network_py <- r_to_py(new_df)

edges_py <- r_to_py(network_df)
```

```{python py-setup}
import pandas as pd, networkx as nx
from pyvis.network import Network
```

```{python py-network}
# Create NetworkX graph object
G = nx.Graph()

# Generate a nested dictionary/tuple with the node of interest ("Symbol") and 
# the relevant attributes (shape and complex)
node_list = []
for i, row in r.sig_network_py.iterrows():
  node_list.append((row["Symbol"], {"color": row.Colors, "Complex": row.Complex, "shape": row.Shapes}))
G.add_nodes_from(node_list)

# Generate a tuple for edges connecting proteins with PTMs
edge_tup = []
for i in range(len(r.edges_py["Symbol"])):
  edge_tup.append((r.edges_py["Symbol"][i], r.edges_py["PTM"][i]))
G.add_edges_from(edge_tup)

# remove self-loops
G.remove_edges_from(nx.selfloop_edges(G))

# now remove nodes w/o edges
lone_nodes = [node for node, degree in G.degree() if degree == 0]
G.remove_nodes_from(lone_nodes)
```

```{python py-network-graph}
# To change the parameters of what's being graphed, adjust the initial call to Network()
net = Network(height="900px", notebook = True, select_menu = True, cdn_resources = "in_line")
net.show_buttons()
net.from_nx(G)
net.save_graph("networkx-pyvis.html")
```

```{r print-HTML}
htmltools::renderDocument(htmltools::htmlTemplate("networkx-pyvis.html"))
```

```{r}
# Identified complexes
final_complexes <- filter(network_df, network_df$Source != "Histone") %>% filter(Complex != "No ID") %>% drop_na() %>% distinct()
final_complex_list <- unique(final_complexes$Complex)

cat(paste("**Networked Complexes:** ", paste(final_complex_list, collapse = ", ")))
```

```{r}
# List members of each
final_members <- data.frame("Complex" = "", "Members" = "", "PTMs" = "")

for (complex in final_complex_list) {
  member_df <- filter(final_complexes, final_complexes$Complex == complex)
  member_list <- paste0(unique(member_df$Symbol), collapse = ", ")
  member_PTMs <- paste0(unique(member_df$PTM), collapse = ", ")
  final_members <- final_members %>% add_row("Complex" = complex, "Members" = member_list, "PTMs" = member_PTMs)
}

knitr::kable(final_members)
```

```{r}
# Differential PTMs
final_PTMs <- distinct(final_complexes, pick(PTM, Diff)) %>% rename("Differential expression" = "Diff")
knitr::kable(final_PTMs)
```


```{r}
# Genes
final_proteins <- unique(final_complexes$Symbol)
final_func <- filter(genes, genes$`HGNC approved symbol` %in% final_proteins) %>% select(!c("UniProt ID (human)"))
knitr::kable(final_func)
```

Separate it out by complex, use table, then make graph manually
